protostar/cheatable_starknet/controllers/expect_events_controller.py

```diff
@@ -0,0 +1,166 @@
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Optional
+
+from protostar.starknet import Address, CairoData, Selector, ReportedException
+
+if TYPE_CHECKING:
+    from protostar.testing import Hook
+    from protostar.cairo_testing import CairoTestExecutionState
+    from protostar.cheatable_starknet.cheatables.cheatable_cached_state import (
+        CheatableCachedState,
+    )
+
+
+@dataclass
+class Event:
+    from_address: Address
+    key: Selector
+    data: Optional[CairoData] = None
+
+
+@dataclass
+class EventMatching:
+    pass
+
+
+@dataclass
+class SkippedEventMatching(EventMatching):
+    emitted_event: Event
+
+
+@dataclass
+class AcceptedEventMatching(EventMatching):
+    emitted_event: Event
+    expected_event: Event
+
+
+@dataclass
+class FailedEventMatching(EventMatching):
+    expected_event: Event
+
+
+@dataclass
+class EventMatchingResult:
+    event_matchings: list[EventMatching]
+
+    @property
+    def failed_event_matchings(self):
+        return [
+            matching
+            for matching in self.event_matchings
+            if isinstance(matching, FailedEventMatching)
+        ]
+
+    @property
+    def should_be_accepted(self) -> bool:
+        return not self.failed_event_matchings
+
+
+class ExpectEventsController:
+    def __init__(
+        self,
+        test_finish_hook: "Hook",
+        test_execution_state: "CairoTestExecutionState",
+        cheatable_state: "CheatableCachedState",
+    ) -> None:
+        self._test_execution_state = test_execution_state
+        self._test_finish_hook = test_finish_hook
+        self._cheatable_state = cheatable_state
+        self._test_finish_hook.on(self.compare_expected_and_actual_results)
+
+    def add_expected_events(self, expected_events: list[Event]):
+        self._test_execution_state.expected_events_list.append(expected_events)
+
+    def compare_expected_and_actual_results(self):
+        for expected_events in self._test_execution_state.expected_events_list:
+            matching_result = match_events(
+                expected_events=expected_events,
+                emitted_events=self._cheatable_state.emitted_events,
+            )
+            if not matching_result.should_be_accepted:
+                raise ExpectEventsMismatchReportedException(matching_result)
+
+
+def match_events(
+    expected_events: list[Event], emitted_events: list[Event]
+) -> EventMatchingResult:
+    event_matchings: list[EventMatching] = []
+    for emitted_event in emitted_events:
+        if not expected_events or not should_accept_event_matching(
```

Comment:
Maybe when this is correct: `if not expected_events`, we could create the `event_matchings` as a list of `SkippedEventMatching` before the loop using comprehension? Would be far cleaner I think.

Reply:
I wouldn't say it cleaner. The loop must be in `else` scope, which would create 3 level nesting. I removed negations from the predicate to make it easier to reason about.

[remove negations](https://github.com/software-mansion/protostar/pull/1412/commits/5a57a99a499d2582c6faa4b6bcad266f741a3cc5)

Reply:
I understand, it looks better now.